### DISCUSSION TITLE
Access input hash in executors?

### DISCUSSION THREAD

**Discussion by wereHamster** [nrwl/nx#12739] [General]
Is there a way to access the input hash in executors? For example to embed the input hash (or more correctly, the hash of all inputs + executor configuration) into the output?

Concretely, I'm interested in using it for the [Next.js Build ID](https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id).

**Comment by minijus** [comment]
`nx print-affected --target...` outputs hashes. It is possible to capture them and use them later.
For example `nx print-affected --target=test --select=tasks.hash` will print comma delimited list of hashes.

**Comment by rfprod** [comment]
@wereHamster
![Screenshot from 2022-11-12 22-19-06](https://user-images.githubusercontent.com/1507770/201493019-76fc32af-d41e-4309-a753-6ac0800f7ecc.png)

```typescript
import { ExecutorContext, Hasher} from '@nrwl/devkit';
```

is this something like what you are looking for?

**Reply by wereHamster** [reply]
Hm, I'd have to look how exactly I can integrate that into my task definition. I currently have something like this:

```json
{
  "targets": {
    "build": {
      "executor": "nx:run-commands",
      "inputs": […],
      "outputs": ["{projectRoot}/.next"],
      "options": {
        "command": "next build"
      }
    }
  }
}
```

Now instead of just `next build` as the command, I need to export a BUILDID environment variable before running `next build`. So something like `BUILDID=$(pnpm nx print-affected --target=build --select=tasks.hash) next build`. Not sure if that's the best approach though…

**Reply by rfprod** [reply]
@wereHamster maybe it's worth switching to a local executor if you're using `nx:run-commands`.
With a local executor, you will have full programmatic control over the thread and will be able to use the `@nrwl/devkit` functions.
https://nx.dev/recipes/executors/creating-custom-executors#local-executors

**Reply by wereHamster** [reply]
Yep, I looked into that. Though I don't see an easy way to generate the hash of all inputs. I saw that there is a `Task` type that contains a `hash`, and I think that's what would be appropriate to use, but that object doesn't seem to be exposed to the executor through the context. Given that `Task` is a «representation of the invocation of an Executor», would it be possible to expose it to the executor?

**Reply by rfprod** [reply]
@wereHamster meybe this will help
there is a `cacheDirectory` configuration in your `nx.json`
![image](https://user-images.githubusercontent.com/1507770/201518454-e2c4c825-6cec-43e6-9a7f-7fc9050b15b2.png)

if you do (use file path relevant to your workspace)
```bash
less /tmp/nx/nx-ng-starter/nxdeps.json
```
you can find something interesting there
![image](https://user-images.githubusercontent.com/1507770/201518555-6407e94e-93fd-4b72-afee-00b9bf01b4e1.png)

I think it shoul be possible to build a task graph in the context of an executor using helpers from `@nrwl/devkit` and `nx`.
You will need to provide them your workspace configuration.
Then, with a high degree of probability, it will be possible to obtain the hash of the task from the graph.
